WordPress/openverse-frontend
# Fix pnpm + docker

## Fixes


Fixes https://github.com/WordPress/openverse-frontend/runs/6283206618

## Description


This PR does the following:

* Pins `pnpm` to a specific version and updates instructions and installations of `pnpm` to indicate how to use the "pinned" version. We need a custom solution because volta does not yet support pinning `pnpm` and we don't use volta to install it in the docker builds anyway.
* Removes the local docker-compose stack for the frontend. It was broken ever since the playwright dockerization and no one noticed, so I'm assuming it was not being used.
* Updates documentation regarding the build
* Removes the multi-stage docker build. Given that the development image is gone now, the multi-stage build gave us nothing. In fact, even before, it wasn't doing anything but making build times longer if you weren't using BuildKit, so good riddance! It's good to delete code.
* Switches to using `pnpm fetch` to populate the store in the docker build from `pnpm-lock.yaml` instead of `pnpm install`. Details on how this works available [on the documentation page for it here](https://pnpm.io/6.x/cli/fetch). I've cross referenced the API and usage with pnpm 7 and there are no differences from pnpm 6, so the "experimental" warning at the top of that documentation page can be safely ignored.
* Updates the production image to run the app using the `node` user rather than root. This is a basic docker security best practice that we've been ignoring for a while. It's a simple fix here, especially without the multi-stage build, so why not.

## Testing Instructions

Checkout this branch and run `docker build . --build-arg PNPM_VERSION=$(cat .pnpm-version) -t openverse-frontend:latest` and confirm that it builds a working image that you can run using `docker run -it -p 127.0.0.1:8443:8443/tcp openverse-frontend:latest` at `localhost:8443`.

## Checklist

- [N/A] I added or updated tests for the changes I made (if applicable).

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
